### PR TITLE
feat: run saber server on 3001 port if 3000 is unavailable

### LINKED ISSUE #182: Run saber server on 3001 port if 3000 is being used
#### What problem does this feature solve?
Running saber server whiteout passing a port argument while the port 3000 is being used break the app 
#### What does the proposed API look like?
Run the server on the port 3001 automatically 
#### How should this be implemented in your opinion?
N/A
#### Are you willing to work on this yourself?
YES

### PR DESCRIPTION
**Pull request by MarchWorks**
**Summary**
Closes #182 
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [ ] Other, please describe:

If changing the UI/CSS related code, please provide the **before/after** screenshot:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)

You have tested in the following browsers: (Providing a detailed version will be better.)

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] IE

If adding a **new feature**, the PR's description includes:

- [ ] A convincing reason for adding this feature
- [ ] Related documents have been updated
- [ ] Related tests have been updated

To avoid wasting your time, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: packages/saber/lib/cli.js [hunk 1]
```diff
@@ -83,6 +73,34 @@ cli.help()
...
+    require('..')(Object.assign({ cwd, dev: true }, options), {
+      server: {
+        host,
+        port
+      },
+      build: {
+        lazy
+      }
+    })
+      .serve()
```
Comment (egoist) [inline packages/saber/lib/cli.js]: we can use https://github.com/sindresorhus/get-port instead, `.serve` has side effects, it runs `.run` before creating the server.

File: packages/saber/lib/cli.js [hunk 2]
```diff
@@ -11,10 +12,16 @@ cli
...
+  .action(async (cwd = '.', options) => {
     setNodeEnv('development')
 
-    const { host, port, lazy } = options
+    let { host, port, lazy } = options
+    // doing the following instead of deleting ignoreOptionDefaultValue: true
+    host = host ? host : '0.0.0.0'
+    // Use the user custom port value even if it's unavailable
+    port = port
+      ? port
```
Comment (MarchWorks) [inline packages/saber/lib/cli.js]: should I use `if` will that look better?

File: packages/saber/lib/cli.js [hunk 3]
```diff
@@ -11,10 +12,15 @@ cli
...
-  .action((cwd = '.', options) => {
+  .action(async (cwd = '.', options) => {
     setNodeEnv('development')
 
-    const { host, port, lazy } = options
+    let { host, port, lazy } = options
+    host = host ? host : '0.0.0.0'
+    if (!port) {
+      port = await getPort({ port: getPort.makeRange(3000, 4000), host })
+    }
```
Comment (egoist) [inline packages/saber/lib/cli.js]: let's validate the `port` in `index.js` instead, because user can also set `port` from `saber-config.js`

**Comment by MarchWorks** [comment]
the failling checks doesn't have anything to do with the PR

File: packages/saber/lib/webpack/PrintStatusPlugin.js [hunk 4]
```diff
@@ -96,6 +96,11 @@ module.exports = class PrintStatusPlugin {
             this.api.config.server.host === '0.0.0.0'
               ? 'localhost'
               : this.api.config.server.host
+          const { port, _originalPort } = this.api.config.server
+          if (port !== _originalPort) {
+            log.warn(`Port ${_originalPort} is in use`)
```
Comment (MarchWorks) [inline packages/saber/lib/webpack/PrintStatusPlugin.js]: port here I think

**Comment by MarchWorks** [comment]
Nice

**Comment by kidonng** [comment]
In a rare case, I run `saber` and it automatically "switched" to a new port (I didn't run any other server). But actually it still serves the pages on port 3000. It doesn't bother a lot so I don't open an issue for this.

```
~\..\..\the-story: master ≣ +1 ~6 -1 | ~5 -0 ! > saber
info Using config file: saber-config.js
info Using theme: ./theme
info Using 20 plugins
success Compiled client successfully in 14s!
info Available at http://localhost:3000
info 165 MB memory used
warning Found a change in your Saber config file, restart server to see the effect.
info (21:3:37) Rebuilding client due to changes in ~\Documents\Projects\the-story\.saber\config.json
success Compiled client successfully in 282ms!
warning Port 3000 is in use, switched to a new port
info Available at http://localhost:3001
info 93.7 MB memory used
info (21:5:4) Rebuilding client due to changes in ~\Documents\Projects\the-story\.saber\config.json
success Compiled client successfully in 257ms!
warning Port 3000 is in use, switched to a new port
info Available at http://localhost:3001
info 98.8 MB memory used
```

**Comment by MarchWorks** [comment]
@kidonng nice catch. found the problem will make a PR to fix it

**Pull request merged**